Add control features

ellwiii opened the pull request:
This is what I have so far, would appreciate any comments and directions:
* for Android Google Maps myLocation button - it does not work without granting location permissions. Requesting location permission seems to be a fairly complicated topic for me, so I am still doing research on it 🥲
* for iOS MapKit myLocation button I've attempted to enable it (also to pick better than default placing and style) but I'm pretty sure it could be done better
* for iOS MapKit levelPicker - there is not much information I've found on this topic, still not clear to me if it is possible to set this control button without having an indoor map implemented in the first place

ellwiii committed:
feature(0004/android/add-control-features): added single props for enabling/disabling UI Controls for Google Maps in Android

Resolve conflicts

ellwiii committed:
feature(0004/android/add-control-features): added UI controls enabling/disabling for Google Maps for Android

Resolve conflicts

ellwiii committed:
Google Maps/ iOS/ added basic control features

Resolve conflicts

ellwiii committed:
MapKit added compass feature

Resolve conflicts

ellwiii committed:
MapKit added my location button feature

Resolve conflicts

ellwiii committed:
MapKit added NSLocationWhenInUseUsageDescription key to Info.plist and location request

ellwiii committed:
MapKit added location request for myLocation button

Resolve conflicts

ellwiii committed:
Google Maps/ iOS/ added current location service to myLocation button

ellwiii committed:
cleanup, added todos

Resolve conflicts

ellwiii committed:
cleanup, added comments

szdziedzic commented on expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt:
```diff
@@ -18,6 +18,22 @@ class ExpoGoogleMapsModule : Module() {
       prop("mapType") { view: GoogleMapsView, mapType: MapType ->
         view.setMapType(mapType)
       }
+      prop("zoomControls") { view: GoogleMapsView, enabled: Boolean ->
+        view.setZoomControlButton(enabled)
+      }
```
What do you think about slightly different and in my opinion more clear naming convention ?
```suggestion
      prop("showZoomControls") { view: GoogleMapsView, enabled: Boolean ->
        view.setShowZoomControlButton(enabled)
      }
```

szdziedzic commented on expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt:
```diff
@@ -18,6 +18,22 @@ class ExpoGoogleMapsModule : Module() {
+      }
+      prop("compass") { view: GoogleMapsView, enabled: Boolean ->
+        view.setCompassButton(enabled)
+      }
```
What do you think about slightly different and in my opinion more clear naming convention ?
```suggestion
      prop("showCompassButton") { view: GoogleMapsView, enabled: Boolean ->
        view.setShowCompassButton(enabled)
      }
```

szdziedzic commented on expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt:
```diff
@@ -18,6 +18,22 @@ class ExpoGoogleMapsModule : Module() {
+      }
+      prop("mapToolbar") { view: GoogleMapsView, enabled: Boolean ->
+        view.setMapToolbarButton(enabled)
+      }
```
```suggestion
      prop("showMapToolbarButton") { view: GoogleMapsView, enabled: Boolean ->
        view.setShowMapToolbarButton(enabled)
      }
```

szdziedzic commented on expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt:
```diff
@@ -18,6 +18,22 @@ class ExpoGoogleMapsModule : Module() {
+        view.setMapToolbarButton(enabled)
+      }
+      //TODO: enable MyLocation layer, otherwise My location button will not be displayed
+      prop("myLocationButton") { view: GoogleMapsView, enabled: Boolean ->
+        view.setMyLocationButton(enabled)
+      }
```
Same question as above

szdziedzic commented on expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt:
```diff
@@ -18,6 +18,22 @@ class ExpoGoogleMapsModule : Module() {
+      prop("myLocationButton") { view: GoogleMapsView, enabled: Boolean ->
+        view.setMyLocationButton(enabled)
+      }
+      prop("levelPicker") { view: GoogleMapsView, enabled: Boolean ->
+        view.setLevelPickerButton(enabled)
+      }
```
Same question as above

szdziedzic commented on expo-maps/build/Maps.types.d.ts:
```diff
@@ -2,11 +2,24 @@ import { ViewProps } from 'react-native';
+    compass: boolean;
+    mapToolbar: boolean;
+    myLocationButton: boolean;
+    levelPicker: boolean;
```
Same question as above

szdziedzic commented on expo-maps/build/Maps.types.d.ts:
```diff
@@ -2,11 +2,24 @@ import { ViewProps } from 'react-native';
     mapType: 'normal' | 'hybrid' | 'satellite' | 'terrain';
+    compass: boolean;
+    myLocationButton: boolean;
+    levelPicker: boolean;
```
Same question as above

szdziedzic commented on expo-maps/build/Maps.types.d.ts:
```diff
@@ -2,11 +2,24 @@ import { ViewProps } from 'react-native';
+    compass?: boolean;
+    mapToolbar?: boolean;
+    myLocationButton?: boolean;
+    levelPicker?: boolean;
```
Same question as above

szdziedzic commented on expo-maps/ios/ExpoMaps/ExpoAppleMapsModule.swift:
```diff
@@ -12,6 +12,15 @@ public class ExpoAppleMapsModule: Module {
       prop("mapType") { (view: AppleMapsView, mapType: MapType) in
         view.setMapType(mapType: mapType)
       }
+      prop("compass") { (view: AppleMapsView, enable: Bool) in
+        view.setCompassButton(enable: enable)
+      }
+      prop("myLocationButton") { (view: AppleMapsView, enable: Bool) in
+        view.setMyLocationButton(enable: enable)
+      }
+      prop("levelPicker") { (view: AppleMapsView, enable: Bool) in
+        view.setLevelPickerButton(enable: enable)
+      }
```
Same question as above

szdziedzic commented on expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt:
```diff
@@ -45,6 +45,46 @@ class GoogleMapsView(context: Context): LinearLayout(context), OnMapReadyCallbac
+      googleMap.uiSettings.isZoomControlsEnabled = enabled
+    }
+  }
+
```
```suggestion
  fun setShowZoomControlButton(enabled: Boolean) {
    updateMap {
      googleMap.uiSettings.isZoomControlsEnabled = enabled
    }
  }

```

szdziedzic commented on expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt:
```diff
@@ -45,6 +45,46 @@ class GoogleMapsView(context: Context): LinearLayout(context), OnMapReadyCallbac
+    updateMap {
+      googleMap.uiSettings.isIndoorLevelPickerEnabled = enabled
+    }
+  }
```
Same as above

szdziedzic commented on expo-maps/ios/ExpoMaps/AppleMapsView.swift:
```diff
@@ -26,5 +26,44 @@ public final class AppleMapsView: UIView {
+    
+  func setLevelPickerButton(enable: Bool) -> Void {
+    //TODO: enable floor picker
+  }
```
Same as above

szdziedzic commented on expo-maps/ios/ExpoMaps/GoogleMapsView.swift:
```diff
@@ -37,4 +37,26 @@ public final class GoogleMapsView: UIView {
+    
+  func setCompassButton(enabled: Bool) -> Void {
+    self.mapView.settings.compassButton = enabled
+  }
```
```suggestion
  func setShowCompassButton(enabled: Bool) -> Void {
    self.mapView.settings.compassButton = enabled
  }
```

szdziedzic commented on expo-maps/ios/ExpoMaps/ExpoGoogleMapsModule.swift:
```diff
@@ -11,6 +11,15 @@ public class ExpoGoogleMapsModule: Module {
       }
+      prop ("compass") { (view: GoogleMapsView, enabled: Bool) in
+        view.setCompassButton(enabled: enabled)
+      }
```
```suggestion
      prop ("showCompassButton") { (view: GoogleMapsView, enabled: Bool) in
        view.setShowCompassButton(enabled: enabled)
      }
```

szdziedzic commented on expo-maps/ios/ExpoMaps/ExpoGoogleMapsModule.swift:
```diff
@@ -11,6 +11,15 @@ public class ExpoGoogleMapsModule: Module {
+      }
+      prop ("levelPicker") { (view: GoogleMapsView, enabled: Bool) in
+        view.setFloorPickerButton(enabled: enabled)
+      }
```
Same question as above

szdziedzic commented on expo-maps/ios/ExpoMaps/GoogleMapsView.swift:
```diff
@@ -37,4 +37,26 @@ public final class GoogleMapsView: UIView {
+  func setFloorPickerButton(enabled: Bool) -> Void {
+      //appears whenever an indoor map is featured prominently
+    self.mapView.settings.indoorPicker = enabled
+  }
```
Same question as above

szdziedzic commented on expo-maps/src/Maps.tsx:
```diff
@@ -10,10 +10,20 @@ export * from './Maps.types';
+  compass: boolean;
+  mapToolbar: boolean;
+  myLocationButton: boolean;
+  levelPicker: boolean;
```
Same as above

szdziedzic commented on expo-maps/src/Maps.types.ts:
```diff
@@ -4,15 +4,28 @@ import { PropsWithChildren } from 'react';
+    compass: boolean;
+    mapToolbar: boolean;
+    myLocationButton: boolean;
+    levelPicker: boolean;
```
Same as above

szdziedzic commented on expo-maps/src/Maps.tsx:
```diff
@@ -10,10 +10,20 @@ export * from './Maps.types';
+  compass: true,
+  mapToolbar: true,
+  myLocationButton: true,
+  levelPicker: true,
```
Same as above

szdziedzic commented on expo-maps/src/Maps.types.ts:
```diff
@@ -4,15 +4,28 @@ import { PropsWithChildren } from 'react';
     mapType: 'normal' | 'hybrid' | 'satellite' | 'terrain';
+    compass: boolean;
+    myLocationButton: boolean;
+    levelPicker: boolean;
```
Same as above

szdziedzic commented on expo-maps/src/Maps.types.ts:
```diff
@@ -4,15 +4,28 @@ import { PropsWithChildren } from 'react';
+    compass?: boolean;
+    mapToolbar?: boolean;
+    myLocationButton?: boolean;
+    levelPicker?: boolean;
```
Same as above

szdziedzic reviewed (commented):
Just a few suggestions

lukmccall commented on expo-maps/ios/ExpoMaps/AppleMapsView.swift:
```diff
@@ -26,5 +26,44 @@ public final class AppleMapsView: UIView {
+        myLocationButton.translatesAutoresizingMaskIntoConstraints = false
+    self.mapView.addSubview(myLocationButton)
+          
+    NSLayoutConstraint.activate([
+        myLocationButton.topAnchor.constraint(equalTo: self.mapView.topAnchor,
+            constant: 100),
+        myLocationButton.trailingAnchor.constraint(equalTo: self.mapView.trailingAnchor, constant: -10)])
+  }
```
```suggestion
    NSLayoutConstraint.activate([
        myLocationButton.topAnchor.constraint(equalTo: self.mapView.topAnchor, constant: 100),
        myLocationButton.trailingAnchor.constraint(equalTo: self.mapView.trailingAnchor, constant: -10)
    ])
  }
```
It looks better in my opinion ;)

lukmccall commented on expo-maps/ios/ExpoMaps/AppleMapsView.swift:
```diff
@@ -26,5 +26,44 @@ public final class AppleMapsView: UIView {
+    let locationManager = CLLocationManager()
+    locationManager.requestWhenInUseAuthorization()
+  }
+
```
```suggestion
```

lukmccall commented on expo-maps/ios/ExpoMaps/AppleMapsView.swift:
```diff
@@ -26,5 +26,44 @@ public final class AppleMapsView: UIView {
       mapViewType = .standard
     }
     self.mapView.mapType = mapViewType
+    
```
```suggestion
```

lukmccall commented on expo-maps/ios/ExpoMaps/AppleMapsView.swift:
```diff
@@ -26,5 +26,44 @@ public final class AppleMapsView: UIView {
+    
   }
+
+  func setCompassButton(enable: Bool) -> Void {
```
```suggestion
  func setCompassButton(enable: Bool) {
```
`Void` type can be omitted.

lukmccall commented on expo-maps/ios/ExpoMaps/AppleMapsView.swift:
```diff
@@ -26,5 +26,44 @@ public final class AppleMapsView: UIView {
+    self.mapView.showsCompass = enable
+  }
+    
+  func setMyLocationButton(enable: Bool) -> Void {
```
```suggestion
  func setMyLocationButton(enable: Bool) {
```
`Void` type can be omitted.

lukmccall commented on expo-maps/ios/ExpoMaps/AppleMapsView.swift:
```diff
@@ -26,5 +26,44 @@ public final class AppleMapsView: UIView {
+    }
+  }
+    
+  func setLevelPickerButton(enable: Bool) -> Void {
```
```suggestion
  func setLevelPickerButton(enable: Bool) {
```
`Void` type can be omitted.

lukmccall commented on expo-maps/ios/ExpoMaps/AppleMapsView.swift:
```diff
@@ -26,5 +26,44 @@ public final class AppleMapsView: UIView {
+    //TODO: enable floor picker
+  }
+    
+  private func enableMyLocationButton() -> Void {
```
```suggestion
  private func enableMyLocationButton() {
```
`Void` type can be omitted.

lukmccall commented on expo-maps/ios/ExpoMaps/AppleMapsView.swift:
```diff
@@ -26,5 +26,44 @@ public final class AppleMapsView: UIView {
+    self.requestLocationPermission()
+
+    self.mapView.showsUserLocation = true
+    let myLocationButton = MKUserTrackingButton(mapView: self.mapView)
+        myLocationButton.layer.backgroundColor = UIColor(white: 1, alpha: 0.5).cgColor
+        myLocationButton.layer.borderColor = UIColor.white.cgColor
+        myLocationButton.layer.borderWidth = 1
+        myLocationButton.layer.cornerRadius = 5
+        myLocationButton.translatesAutoresizingMaskIntoConstraints = false
```
```suggestion
    let myLocationButton = MKUserTrackingButton(mapView: self.mapView)
    myLocationButton.layer.backgroundColor = UIColor(white: 1, alpha: 0.5).cgColor
    myLocationButton.layer.borderColor = UIColor.white.cgColor
    myLocationButton.layer.borderWidth = 1
    myLocationButton.layer.cornerRadius = 5
    myLocationButton.translatesAutoresizingMaskIntoConstraints = false
```

lukmccall commented on expo-maps/ios/ExpoMaps/AppleMapsView.swift:
```diff
@@ -26,5 +26,44 @@ public final class AppleMapsView: UIView {
+        myLocationButton.trailingAnchor.constraint(equalTo: self.mapView.trailingAnchor, constant: -10)])
+  }
+    
+  private func requestLocationPermission() -> Void {
```
```suggestion
  private func requestLocationPermission() {
```

lukmccall commented on expo-maps/ios/ExpoMaps/GoogleMapsView.swift:
```diff
@@ -37,4 +37,26 @@ public final class GoogleMapsView: UIView {
+  }
+
+  func setFloorPickerButton(enabled: Bool) -> Void {
+      //appears whenever an indoor map is featured prominently
```
```suggestion
    //appears whenever an indoor map is featured prominently
```

lukmccall commented on expo-maps/ios/ExpoMaps/GoogleMapsView.swift:
```diff
@@ -37,4 +37,26 @@ public final class GoogleMapsView: UIView {
+  func setMyLocationButton(enabled: Bool) -> Void {
+    self.mapView.settings.myLocationButton = enabled
+    if (enabled == true) {
+        self.requestLocationPermission()
+        self.mapView.isMyLocationEnabled = true
```
```suggestion
      self.requestLocationPermission()
      self.mapView.isMyLocationEnabled = true
```

lukmccall reviewed (commented):
Looks good to me 🏅 
Just a couple of comments: 
- I would choose the naming proposed by Szymon.
- You don't have to ask for location permission manually. Please install `expo-location` (https://github.com/expo/expo/tree/master/packages/expo-location#installation-in-bare-react-native-projects) and call `Location.requestForegroundPermissions` via js. We can add a native permission requester later.

ellwiii committed:
cleanup, resolved added comments

Resolve conflicts

szdziedzic commented on expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt:
```diff
@@ -45,6 +45,46 @@ class GoogleMapsView(context: Context): LinearLayout(context), OnMapReadyCallbac
+      googleMap.uiSettings.isIndoorLevelPickerEnabled = enabled
+    }
+  }
+
```
I'm wondering if it wouldn't be more consistent solution to move this logic into `GoogleMapsControls `class as I did in my [markers PR](https://github.com/software-mansion-labs/expo-maps/pull/17) or @Zwatotem did in his PR for gestures. What's your opinion?

szdziedzic commented on expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt:
```diff
@@ -18,6 +18,22 @@ class ExpoGoogleMapsModule : Module() {
         view.setMapType(mapType)
       }
+      prop("showZoomControls") { view: GoogleMapsView, enabled: Boolean ->
+        view.setShowZoomControlButton(enabled)
```
```suggestion
        view.setShowZoomControls(enabled)
```

Maybe we should try to rename functions to match prop names. What do you think? Overall everything looks good here, it's just cosmetic change suggestion.

szdziedzic commented on expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt:
```diff
@@ -18,6 +18,22 @@ class ExpoGoogleMapsModule : Module() {
+        view.setShowZoomControlButton(enabled)
+      }
+      prop("showCompass") { view: GoogleMapsView, enabled: Boolean ->
+        view.setShowCompassButton(enabled)
```
As above

szdziedzic commented on expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt:
```diff
@@ -18,6 +18,22 @@ class ExpoGoogleMapsModule : Module() {
+        view.setShowCompassButton(enabled)
+      }
+      prop("showMapToolbar") { view: GoogleMapsView, enabled: Boolean ->
+        view.setShowMapToolbarButton(enabled)
```
As above

szdziedzic commented on expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt:
```diff
@@ -18,6 +18,22 @@ class ExpoGoogleMapsModule : Module() {
+      }
+      //TODO: enable MyLocation layer, otherwise My location button will not be displayed
+      prop("showMyLocationButton") { view: GoogleMapsView, enabled: Boolean ->
+        view.setShowMyLocationButton(enabled)
```
As above

szdziedzic commented on expo-maps/android/src/main/java/expo/modules/maps/ExpoGoogleMapsModule.kt:
```diff
@@ -18,6 +18,22 @@ class ExpoGoogleMapsModule : Module() {
+        view.setShowMyLocationButton(enabled)
+      }
+      prop("showLevelPicker") { view: GoogleMapsView, enabled: Boolean ->
+        view.setShowLevelPickerButton(enabled)
```
As above

szdziedzic commented on expo-maps/ios/ExpoMaps/AppleMapsView.swift:
```diff
@@ -27,4 +27,43 @@ public final class AppleMapsView: UIView {
   }
+
+  func setShowCompassButton(enable: Bool) {
+    self.mapView.showsCompass = enable
+  }
+    
+  func setShowMyLocationButton(enable: Bool) {
+    if (enable == true) {
+      self.enableMyLocationButton()
+    }
+  }
+    
+  func setShowLevelPickerButton(enable: Bool) {
+    //TODO: enable floor picker
+  }
+    
+  private func enableMyLocationButton() {
+    self.requestLocationPermission()
+
+    self.mapView.showsUserLocation = true
+    let myLocationButton = MKUserTrackingButton(mapView: self.mapView)
+    myLocationButton.layer.backgroundColor = UIColor(white: 1, alpha: 0.5).cgColor
+    myLocationButton.layer.borderColor = UIColor.white.cgColor
+    myLocationButton.layer.borderWidth = 1
+    myLocationButton.layer.cornerRadius = 5
+    myLocationButton.translatesAutoresizingMaskIntoConstraints = false
+    self.mapView.addSubview(myLocationButton)
+          
+    NSLayoutConstraint.activate([
+        myLocationButton.topAnchor.constraint(equalTo: self.mapView.topAnchor,
+constant: 100),
+        myLocationButton.trailingAnchor.constraint(equalTo:
+self.mapView.trailingAnchor, constant: -10)
+    ])
+  }
+    
+  private func requestLocationPermission() {
+    let locationManager = CLLocationManager()
+    locationManager.requestWhenInUseAuthorization()
+  }
```
As above, we can try to move this logic to `AppleMapsControls` class

szdziedzic commented on expo-maps/ios/ExpoMaps/ExpoAppleMapsModule.swift:
```diff
@@ -12,6 +12,15 @@ public class ExpoAppleMapsModule: Module {
       prop("mapType") { (view: AppleMapsView, mapType: MapType) in
         view.setMapType(mapType: mapType)
       }
+      prop("showCmpass") { (view: AppleMapsView, enable: Bool) in
```
```suggestion
      prop("showCompass") { (view: AppleMapsView, enable: Bool) in
```

szdziedzic replied:
```suggestion
      prop("showCompass") { (view: AppleMapsView, enable: Bool) in
```

I think this is typo

szdziedzic commented on expo-maps/ios/ExpoMaps/ExpoAppleMapsModule.swift:
```diff
@@ -12,6 +12,15 @@ public class ExpoAppleMapsModule: Module {
         view.setMapType(mapType: mapType)
       }
+      prop("showCmpass") { (view: AppleMapsView, enable: Bool) in
+        view.setShowCompassButton(enable: enable)
```
Same suggestion as for Android
```suggestion
        view.setShowCompass(enable: enable)
```

szdziedzic commented on expo-maps/ios/ExpoMaps/ExpoAppleMapsModule.swift:
```diff
@@ -12,6 +12,15 @@ public class ExpoAppleMapsModule: Module {
+        view.setShowCompassButton(enable: enable)
+      }
+      prop("showMyLocationButton") { (view: AppleMapsView, enable: Bool) in
+        view.setShowMyLocationButton(enable: enable)
```
As above

szdziedzic commented on expo-maps/ios/ExpoMaps/ExpoAppleMapsModule.swift:
```diff
@@ -12,6 +12,15 @@ public class ExpoAppleMapsModule: Module {
+        view.setShowMyLocationButton(enable: enable)
+      }
+      prop("showLevelPicker") { (view: AppleMapsView, enable: Bool) in
+        view.setShowLevelPickerButton(enable: enable)
```
As above

szdziedzic commented on expo-maps/ios/ExpoMaps/GoogleMapsView.swift:
```diff
@@ -37,4 +37,26 @@ public final class GoogleMapsView: UIView {
     }
     self.mapView.mapType = mapViewType
   }
+    
+  func setShowCompassButton(enabled: Bool) {
+    self.mapView.settings.compassButton = enabled
+  }
+
+  func setShowMyLocationButton(enabled: Bool) {
+    self.mapView.settings.myLocationButton = enabled
+    if (enabled == true) {
+     self.requestLocationPermission()
+     self.mapView.isMyLocationEnabled = true
+    }
+  }
+
+  func setShowFloorPickerButton(enabled: Bool) {
+    //appears whenever an indoor map is featured prominently
+    self.mapView.settings.indoorPicker = enabled
+  }
+    
+    private func requestLocationPermission() {
+      let locationManager = CLLocationManager()
+      locationManager.requestWhenInUseAuthorization()
+    }
```
We can also try to move this logic to `GoogleMapsControls` class

szdziedzic reviewed (commented):
Good job 🚀! Everything looks correct 🎉! I'm so sorry about being so nitpicky, but I left few more cosmetic suggestions 😅

ellwiii committed:
Adding iOS controlls classes, naming cleanup

Resolve conflicts

ellwiii committed:
Adding Android controllers class, naming cleanup

Resolve conflicts

ellwiii committed:
Adding configurations for expo-location

Resolve conflicts

ellwiii commented:
Update from me:
I have applied changes proposed by @szdziedzic:

- changed functions naming to mach pops names
- unified "enable" parameters name
- moved logic of setting controllers into new classes

About expo-location setup - I've not yet figured out the correct way to use it. I was trying to prompt user for allowing location permissions via calling requestForegroundPermissionsAsync() in App.js but it didn't provide the expected result. I will try to resolve this issue in the following days.

ellwiii committed:
Enabled getting location permissions via expo-location

Resolve conflicts

ellwiii commented:
I enabled getting location permissions from users via expo-location (the previous problem turned out to be you have to manually reset location permissions settings on iOS simulator and Android emulator - that's why the "ask for permissions" window was not visible to me). As @lukmccall recommended, this solution also in my opinion is fine for now ☺️

Zwatotem reviewed (commented):
I don't see any protocol added to bond GoogleMapsControllers and AppleMapsControlles with a common interface, like it is done on the Android side. This seems counter-intuitive to have an interface, where there is one implementation and no interface, when there's multiple.

ellwiii committed:
Added protocol with common interface for AppleMapsControllers and GoogleMapsControllers

ellwiii commented:
> I don't see any protocol added to bond GoogleMapsControllers and AppleMapsControlles with a common interface, like it is done on the Android side. This seems counter-intuitive to have an interface, where there is one implementation and no interface, when there's multiple.

You are completely right @Zwatotem, I have omitted this in my implementation. Now it's added.

szdziedzic commented on expo-maps/android/src/main/java/expo/modules/maps/Controllers.kt:
```diff
@@ -0,0 +1,9 @@
+    fun setShowMapToolbar(enable: Boolean)
+    fun setShowMyLocationButton(enable: Boolean)
+    fun setShowLevelPicker(enable: Boolean)
+}
```
```suggestion
}

```

szdziedzic commented on example/App.tsx:
```diff
@@ -1,10 +1,22 @@
 
 export default function App() {
+
+  const getLocationPermissions = async () => {
+    await Location.requestForegroundPermissionsAsync();
+  }
+
+  useEffect(() => {
+    getLocationPermissions();
+  }, []);
```
I'm wondering if user should do this in his App before using `Expo Maps ` or maybe if we should do this as a part of library functionalities. What do you think?

ellwiii replied:
@szdziedzic This is only a temporary solution. I believe finally it should be a part of library functionalities.

In my opinion we should create separate issue to work on this topic. The solution will also depend strongly whether we'd like to keep using expo-location or switch to native ways of getting location permissions from user.

szdziedzic replied:
Sounds reasonable 👍

szdziedzic reviewed (approved):
Nice one. Looks alright 👍. It's good to go in my opinion 👏 . I leave few last comments for you:

ellwiii committed:
Clean up: fixed typos, moved swift files to correct folder.

Resolve conflicts

ellwiii committed:
Clean up

Resolve conflicts

ellwiii committed:
Clean up: renaming files
